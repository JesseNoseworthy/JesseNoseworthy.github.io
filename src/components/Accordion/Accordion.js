import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styled from 'styled-components';

const AccordionContainer = styled.div`width: 100%;`;

class Accordion extends Component {
  render() {
    const { children, isAllOpen } = this.props;
    return (
      <AccordionContainer>
        {React.Children.map(children, child =>
          React.cloneElement(child, { isAllOpen })
        )}
      </AccordionContainer>
    );
  }
}

Accordion.propTypes = {
  children: PropTypes.any,
  isAllOpen: PropTypes.bool,
};

export default Accordion;
