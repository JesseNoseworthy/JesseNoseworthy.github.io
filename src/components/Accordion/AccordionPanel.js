import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styled from 'styled-components';

const AccordionPanelContainer = styled.div`width: 100%;`;

class AccordionPanel extends Component {
  constructor(props) {
    super(props);
    this.state = { isOpen: this.props.isOpen || this.props.isOpenOnLoad || this.props.isAllOpen };
    this.togglePanel = this.togglePanel.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.isAllOpen !== undefined && nextProps.isAllOpen !== null && this.props.isAllOpen !== nextProps.isAllOpen) {
      this.setState({ isOpen: nextProps.isAllOpen });
    } else {
      if (!this.props.isOpen && nextProps.isOpen) {
        this.setState({ isOpen: nextProps.isOpen });
      }
    }
  }

  togglePanel() {
    if (this.props.onToggle) this.props.onToggle(!this.state.isOpen);
    this.setState({ isOpen: !this.state.isOpen });
  }

  render() {
    return (
      <AccordionPanelContainer className={this.props.className}>
        {React.Children.map(this.props.children, child =>
          React.cloneElement(child, {
            isOpen: this.state.isOpen,
            togglePanel: this.togglePanel,
          })
        )}
      </AccordionPanelContainer>
    );
  }
}

AccordionPanel.propTypes = {
  isOpen: PropTypes.bool,
  isOpenOnLoad: PropTypes.bool,
  isAllOpen: PropTypes.bool,
  children: PropTypes.any,
  onToggle: PropTypes.func,
}

export default AccordionPanel;
