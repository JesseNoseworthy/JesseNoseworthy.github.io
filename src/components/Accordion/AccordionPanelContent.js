import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styled from 'styled-components';
import media from 'utils/media';

const AccordionPanelContentContainer = styled.div`
  max-height: ${props => (props.isOpen ? `${props.maxHeight}px` : 0)};
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  visibility: ${props => (props.isOpen ? 'visible' : 'hidden')};
  transition: max-height 500ms, visibility 500ms;
  margin: 15px 0;
`;

class AccordionPanelContent extends Component {
  constructor(props) {
    super(props);

    this.state = {
      maxHeight: 0,
    };

    this.getSectionHeight = this.getSectionHeight.bind(this);
  }

  componentDidMount() {
    this.getSectionHeight();
  }

  componentWillReceiveProps(nextProps) {
    if (
      nextProps.children !== this.props.children ||
      nextProps.isOpen !== this.props.isOpen
    ) {
      this.getSectionHeight();
    }
  }

  getSectionHeight() {
    if (this.container) {
      this.setState({
        maxHeight: this.container.scrollHeight,
      });
    }
  }

  render() {
    const { isOpen, children, className } = this.props;
    return (
      <AccordionPanelContentContainer
        isOpen={isOpen}
        className={className}
        maxHeight={this.state.maxHeight}
        innerRef={el => (this.container = el)}
      >
        {children}
      </AccordionPanelContentContainer>
    );
  }
}

AccordionPanelContent.propTypes = {
  isOpen: PropTypes.bool,
  children: PropTypes.any,
  className: PropTypes.string,
};

export default AccordionPanelContent;
