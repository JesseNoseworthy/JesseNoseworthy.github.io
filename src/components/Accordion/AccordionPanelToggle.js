import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styled from 'styled-components';

const AccordionPanelToggleContainer = styled.div`
  cursor: ${props => props.isDisabled ? 'auto' : 'pointer'};
`;

class AccordionPanelToggle extends Component {
  constructor(props) {
    super(props);

    this.handleSelect = this.handleSelect.bind(this);
  }

  handleSelect() {
    if (!this.props.isDisabled && this.props.togglePanel) this.props.togglePanel();
  }

  render() {
    const { isOpen, children, className, isDisabled } = this.props;
    return (
      <AccordionPanelToggleContainer className={className} isOpen={isOpen} onClick={() => this.handleSelect()} isDisabled={isDisabled}>
        {React.Children.map(children, (child) =>
          React.cloneElement(child, { isOpen })
        )}
      </AccordionPanelToggleContainer>
    );
  }
}

AccordionPanelToggle.propTypes = {
  isOpen: PropTypes.bool,
  togglePanel: PropTypes.func,
  children: PropTypes.any,
  className: PropTypes.string,
  isDisabled: PropTypes.bool,
}

export default AccordionPanelToggle;
