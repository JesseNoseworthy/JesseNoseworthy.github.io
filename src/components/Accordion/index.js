import Accordion from './Accordion';
import AccordionPanel from './AccordionPanel';
import AccordionPanelToggle from './AccordionPanelToggle';
import AccordionPanelContent from './AccordionPanelContent';

module.exports = {
  Accordion,
  AccordionPanel,
  AccordionPanelToggle,
  AccordionPanelContent
}
