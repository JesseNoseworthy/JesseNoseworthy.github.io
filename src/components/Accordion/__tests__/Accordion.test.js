import React from 'react';
import renderer from 'react-test-renderer';
import Accordion from '../Accordion';

describe('Snapshot tests', () => {
  it('default state', () => {
    const component = renderer.create(<Accordion />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
