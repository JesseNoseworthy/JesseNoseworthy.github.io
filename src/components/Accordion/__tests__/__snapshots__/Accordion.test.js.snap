// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests default state 1`] = `
<div
  className="Accordion__AccordionContainer-lhzlfC cTooy"
/>
`;
